### PR TITLE
Add update-tengo-map subcommand, Tengo map utilities, tests, and docs

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces a new stilyagi subcommand: `update-tengo-map` to merge entries from a source list into a named Tengo map (default map: `allow`).
- Adds a generic Tengo map utility module to parse sources and update Tengo scripts in place while preserving existing entries.
- Includes unit and behavior tests, plus updated documentation.

## Changes
- New module: concordat_vale/tengo_map.py
  - Implemented functionalities:
    - MapValueType: supports TRUE, STRING, BOOLEAN, NUMBER parsing modes.
    - parse_source_entries(source, value_type): reads and parses source lists, stripping comments and blank lines.
    - update_tengo_map(tengo_path, map_name, entries): updates or appends map entries inside a Tengo script, preserving existing non-mentioned entries.
    - Internal helpers and _Entry data structure to track line indices, indentation, and comments.
  - Added TengoMapError for error handling.
  - Exposed API via __all__.
- CLI integration: concordat_vale/stilyagi.py
  - Added update_tengo_map_command with parameters:
    - --project-root (Path): project root for resolving paths
    - --source (Path): path to the source entries file
    - --dest (str): Tengo script path with optional ::mapname (default map is `allow`)
    - --type (str): value parsing mode: `true`, `=`, `=b`, `=n` (default `true`)
  - Behavior: resolves paths, parses source entries, updates the Tengo map in place, and prints a concise summary like
    `2 entries provided, 2 updated`.
  - Uses internal helpers: `_split_dest`, `_coerce_value_type`, `_read_pyproject_version` for validation and conversion.
- Tests
  - Unit tests: tests/test_tengo_map.py covering parsing, numeric/string/boolean handling, map updates, and error cases.
  - Behaviour tests: added features/stilyagi_update_tengo_map.feature and tests/behaviour/test_stilyagi_update_tengo_map_steps.py to verify CLI behavior across scenarios.
- Documentation
  - docs/stilyagi-design.md: describes new `update-tengo-map` sub-command and its behavior.
  - docs/users-guide.md: new section “Updating Tengo maps with stilyagi” detailing usage, sources, destinations, and value types.
  - docs/local-validation-of-github-actions-with-act-and-pytest.md: minor wording/style tweaks for clarity.
- Tests and minor fixes
  - Small test assertion formatting improvements in tests/test_stilyagi.py and tests/test_vale_packages.py to improve determinism and readability.
  - test_helpers/valedate.py adjusted to ensure correct typing when materialising styles, improving type safety in tests.

## Why
- Provides a robust, generic mechanism to merge entries into Tengo maps directly from a source list, reducing manual editing and merge conflicts.
- Keeps the Tengo map editing in-place while respecting existing entries not mentioned in the source.
- Adds flexible value parsing (true, string, boolean, numeric) to cover common use cases.

## How to use
- Basic usage (default map `allow`):
  - stilyagi update-tengo-map --source path/to/entries.txt --dest path/to/script.tengo
- Targeting a named map:
  - stilyagi update-tengo-map --source path/to/entries.txt --dest path/to/script.tengo::exceptions
- Change value parsing, e.g. numeric values:
  - stilyagi update-tengo-map --source path/to/entries.txt --dest path/to/script.tengo::exceptions --type =n
- Destination suffix defaults to the `allow` map if omitted.

## Test plan
- Unit tests for parsing, updating, and error cases pass locally.
- Behavioural tests verify:
  - Boolean entries get merged into the default map.
  - Numeric values override or add entries in a named map and summarize updates correctly.
- Documentation builds should reflect the new command and usage.

## Known limitations / Next steps
- If you later want to merge multiple maps or support more complex write strategies (e.g., comments preservation per-entry), we can extend the parser and renderer accordingly.
- Additional integration tests could exercise more edge cases (e.g., complex values, nested comments).

🌿 Generated by [Terry](https://www.terragonlabs.com)

---

ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/aeed6a4b-0a5f-4b6e-94e0-045a12b92999

## Summary by Sourcery

Introduce a generic Tengo map update utility and integrate it into the stilyagi CLI via a new update-tengo-map subcommand.

New Features:
- Add a generic tengo_map module to parse entry lists and update Tengo map literals in-place while preserving existing entries.
- Expose a stilyagi update-tengo-map CLI subcommand to merge source entries into a specified Tengo map with configurable value parsing modes.

Enhancements:
- Refactor Vale test helpers to share common command execution logic and improve test readability and reuse.
- Tighten typing in the test style materialisation helper to ensure styles mappings are handled safely.

Documentation:
- Document the new stilyagi update-tengo-map subcommand and its usage in the user guide and design docs, and make minor wording tweaks to existing docs.

Tests:
- Add unit tests for the tengo_map utilities covering parsing modes, map updates, and error handling.
- Add BDD-style behavioural tests for the stilyagi update-tengo-map CLI to validate integration with Tengo scripts and reported summaries.
- Tidy existing tests for stilyagi and Vale packages to improve determinism and clarity of assertions.
</pr-body>

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[concordat_vale/tengo_map.py](https://github.com/leynos/concordat-vale/pull/8/files/9a4d5c79b3e42f020bc6ab034d130af23eb4dca6#diff-dc5cdec2cd49fa1e7963a7688eae44d7ff739b82bc3323303c76da098185fb74)

Comment on lines +175 to +198

~~~python
def _parse_token(token: str, value_type: MapValueType) -> tuple[str, object]:
    if value_type is MapValueType.TRUE:
        return token, True

    if "=" not in token:
        msg = "Source lines must include '=' when using typed modes."
        raise TengoMapError(msg)

    key, raw_value = token.split("=", 1)
    key = key.strip()
    value = raw_value.strip()
    if not key:
        msg = "Map keys may not be empty."
        raise TengoMapError(msg)

    if value_type is MapValueType.STRING:
        return key, _parse_string_value(value)
    if value_type is MapValueType.BOOLEAN:
        return key, _parse_boolean_value(value)
    if value_type is MapValueType.NUMBER:
        return key, _parse_numeric_value(value)

    msg = f"Unsupported map value type: {value_type}"
    raise TengoMapError(msg)
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/73479/delta?repo-id=1240187&review-id=8&biomarker=Complex+Method&filename=concordat_vale%2Ftengo_map.py&method=_parse_token)
_parse_token has a cyclomatic complexity of 10, threshold = 9

File: concordat_vale/tengo_map.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+import re
+import typing as typ
+
+if typ.TYPE_CHECKING:
+    import collections.abc as cabc
+    from pathlib import Path
+
+ENTRY_PATTERN = re.compile(
+    r'^(?P<indent>\s*)"(?P<key>[^"\\]+)"\s*:\s*(?P<value>[^,]+),'
+    r"(?P<comment>\s*//.*)?\s*$"
```
[the remark this answers is not in the document]
Comment (leynos) [reply concordat_vale/tengo_map.py]: Has this now been resolved?

File: tests/behaviour/test_stilyagi_update_tengo_map_steps.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    scenario_state["stdout"] = stdout_lines[-1] if stdout_lines else ""
+    return result
+
+
+@then("the allow map contains the boolean entries")
+def allow_map_contains_entries(scenario_state: ScenarioState) -> None:
+    """Verify that the allow map was updated."""
+    contents = scenario_state["tengo_path"].read_text(encoding="utf-8")
+    assert '"ALPHA": true,' in contents
+    assert '"BETA": true,' in contents
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/behaviour/test_stilyagi_update_tengo_map_steps.py]: Has this now been resolved?

File: concordat_vale/tengo_map.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+            continue
+
+        entries_provided += 1
+        key, value = _parse_token(token, value_type)
+        parsed[key] = value
+
+    return entries_provided, parsed
+
+
+def update_tengo_map(
```
[the remark this answers is not in the document]
Comment (leynos) [reply concordat_vale/tengo_map.py]: Has this now been resolved?

File: concordat_vale/tengo_map.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+def _values_equal(existing: object, new_value: object) -> bool:
+    if isinstance(existing, (int, float)) and isinstance(new_value, (int, float)):
+        return float(existing) == float(new_value)
+    return existing == new_value
+
+
+def _render_entry(key: str, value: object, indent: str, comment: str) -> str:
+    rendered_value = _render_value(value)
+    suffix = comment if comment else ""
```
[the remark this answers is not in the document]
Comment (leynos) [reply concordat_vale/tengo_map.py]: Has this now been resolved?

File: concordat_vale/tengo_map.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    return stripped
+
+
+def _try_parse_boolean(raw: str) -> bool | None:
+    lowered = raw.lower()
+    if lowered == "true":
+        return True
+    if lowered == "false":
+        return False
+    return None
```
[the remark this answers is not in the document]
Comment (leynos) [reply concordat_vale/tengo_map.py]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[concordat_vale/tengo_map.py](https://github.com/leynos/concordat-vale/pull/8/files/cb8ede36750d68443f32edf08e0f7e6b7decfd92#diff-dc5cdec2cd49fa1e7963a7688eae44d7ff739b82bc3323303c76da098185fb74)

Comment on lines +194 to +217

~~~python
def _apply_single_entry(
    lines: list[str],
    existing: dict[str, _Entry],
    key: str,
    value: object,
    entry_indent: str,
    closing_idx: int,
) -> tuple[int, int]:
    """Apply an update for a single key, returning delta-updated and new index."""
    if key in existing:
        entry = existing[key]
        if _values_equal(entry.value, value):
            return 0, closing_idx
        lines[entry.index] = _render_entry(
            key=key,
            value=value,
            indent=entry.indent,
            comment=entry.comment,
        )
        return 1, closing_idx

    rendered_line = _render_entry(key, value, entry_indent, "")
    lines.insert(closing_idx, rendered_line)
    return 1, closing_idx + 1
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/73479/delta?repo-id=1240187&review-id=8&biomarker=Excess+Number+of+Function+Arguments&filename=concordat_vale%2Ftengo_map.py&method=_apply_single_entry)
_apply_single_entry has 6 arguments, max arguments = 4

File: concordat_vale/tengo_map.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    )
+    updated = _process_entries(ctx, entries)
+    lines = ctx.lines
+
+    new_text = "\n".join(lines) + "\n"
+    wrote_file = new_text != text
+    if wrote_file:
+        tengo_path.write_text(new_text, encoding="utf-8")
+    return MapUpdateResult(updated=updated, wrote_file=wrote_file)
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply concordat_vale/tengo_map.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: tests/behaviour/test_stilyagi_update_tengo_map_steps.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        "-m",
+        "concordat_vale.stilyagi",
+        "update-tengo-map",
+        "--project-root",
+        str(project_root),
+        str(source_path),
+        dest_argument,
+        *extra_args,
+    ]
+    result = subprocess.run(  # noqa: S603, TD001  # FIXME(@assistant): false positive for S603; controlled arg list in tests; see https://github.com/leynos/concordat-vale/issues/999
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/behaviour/test_stilyagi_update_tengo_map_steps.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.
